File: ports/lua/src/unicode_lua/security/display/filename_disguise.lua
-- FilenameDisguise — detection of filename/extension disguise attacks where the
-- visible extension differs from the byte extension (the display-layer
-- detector).
--
-- Byte-faithful transcription of the verified Rust reference implementation,
-- itself a transcription of `Unicode/Security/Display/FilenameDisguise.lean`.
--
-- Threat model. An adversary delivers a file whose rendered name looks like a
-- benign type (`document.txt`) but whose actual byte extension is executable —
-- the canonical attack inserts U+202E RIGHT-TO-LEFT OVERRIDE so
-- `document<RLO>txt.exe` renders as `document exe.txt`.
--
-- What the detector draws. Detection is presentation- and language-agnostic: it
-- surfaces every codepoint that could cause display-vs-byte divergence in the
-- filename — any bidi format-control anywhere, and any fullwidth/halfwidth or
-- combining (grapheme Extend) codepoint in the extension region (after the last
-- `.`). Native-RTL names with no bidi controls clear. It reuses the port's own
-- tables (the BidiControlBalance format-control set, the grapheme-segmentation
-- Extend class, the inlined fullwidth range), never a host filesystem or
-- rendering library.
--
-- Sub-threats (priority order):
--   1. RloFlip            any bidi format-control in the input.
--   2. WidthClassExt      a fullwidth/halfwidth codepoint in the extension.
--   3. CombiningInExt     a combining (Extend) codepoint in the extension.
--   4. MultipleExtensions >= 3 dots (advisory; e.g. legitimate `.tar.gz.sig`).
--
-- Codepoint positions are 0-based to mirror the reference.

local bidi = require("unicode_lua.security.covert.bidi_control_balance")
local grapheme = require("unicode_lua.segmentation.grapheme")

local unpack = table.unpack or unpack

local M = {}

-- ─────────────────────────────────────────────────────────────────────
-- §1 Constants
-- ─────────────────────────────────────────────────────────────────────

-- The count of `.` separators at or beyond which the name is treated as a
-- multiple-extension advisory hazard.
M.MIN_MULTI_EXT = 3

-- ─────────────────────────────────────────────────────────────────────
-- §2 Core predicates (reuse the port's own tables)
-- ─────────────────────────────────────────────────────────────────────

-- True iff `cp` is U+002E FULL STOP (the extension separator).
function M.is_ascii_dot(cp)
  return cp == 0x002E
end

-- True iff `cp` is in the Halfwidth and Fullwidth Forms block.
function M.is_fullwidth_halfwidth(cp)
  return cp >= 0xFF01 and cp <= 0xFFEF
end

-- True iff `cp` is a bidi format-control (reuses the BidiControlBalance
-- predicate over the LRE/RLE/LRO/RLO/PDF/LRI/RLI/FSI/PDI set).
function M.is_bidi_format_control(cp)
  return bidi.is_bidi_format_control(cp)
end

-- True iff `cp` has Grapheme_Cluster_Break = Extend (reuses the grapheme-
-- segmentation GCB table via `lookup_gcb`, exactly as the Rust reference does).
function M.is_grapheme_extend(cp)
  return grapheme.lookup_gcb(cp) == "Extend"
end

-- ─────────────────────────────────────────────────────────────────────
-- §3 Sub-detectors
-- ─────────────────────────────────────────────────────────────────────

-- 0-based positions of every `.` in `input`.
local function dot_positions(input)
  local dots = {}
  for i = 1, #input do
    if M.is_ascii_dot(input[i]) then
      dots[#dots + 1] = i - 1
    end
  end
  return dots
end

-- Position (0-based) and codepoint of the first bidi format-control, or nil.
local function first_bidi_control(input)
  for i = 1, #input do
    if M.is_bidi_format_control(input[i]) then
      return i - 1, input[i]
    end
  end
  return nil, nil
end

-- Position (0-based) and codepoint of the first fullwidth/halfwidth codepoint
-- at or after the 0-based `start`, or nil.
local function first_fullwidth_from(input, start)
  for i = 1, #input do
    local idx = i - 1
    if idx >= start and M.is_fullwidth_halfwidth(input[i]) then
      return idx, input[i]
    end
  end
  return nil, nil
end

-- Position (0-based) and codepoint of the first Extend codepoint at or after
-- the 0-based `start`, or nil.
local function first_extend_from(input, start)
  for i = 1, #input do
    local idx = i - 1
    if idx >= start and M.is_grapheme_extend(input[i]) then
      return idx, input[i]
    end
  end
  return nil, nil
end

-- Count of fullwidth/halfwidth codepoints at or after the 0-based `start`.
local function count_fullwidth_from(input, start)
  local count = 0
  for i = 1, #input do
    if (i - 1) >= start and M.is_fullwidth_halfwidth(input[i]) then
      count = count + 1
    end
  end
  return count
end

-- Count of Extend codepoints at or after the 0-based `start`.
local function count_extend_from(input, start)
  local count = 0
  for i = 1, #input do
    if (i - 1) >= start and M.is_grapheme_extend(input[i]) then
      count = count + 1
    end
  end
  return count
end

-- Count of bidi format-controls anywhere in `input`.
local function count_bidi_control(input)
  local count = 0
  for i = 1, #input do
    if M.is_bidi_format_control(input[i]) then
      count = count + 1
    end
  end
  return count
end

-- ─────────────────────────────────────────────────────────────────────
-- §4 Top-level detection
-- ─────────────────────────────────────────────────────────────────────

-- The FilenameDisguise detection function. Returns a verdict table mirroring
-- the Rust `Verdict`: `input`, `classify` (`{ kind, sub, positions, decoded }`),
-- `dot_positions`, `last_dot_pos`, `bidi_control_count`, `fullwidth_in_ext`,
-- `combining_in_ext`.
function M.detect(input)
  local dots = dot_positions(input)
  local last_dot = nil
  if #dots > 0 then
    last_dot = dots[#dots]
  end
  local ext_start
  if last_dot ~= nil then
    ext_start = last_dot + 1
  else
    ext_start = #input
  end

  local bidi_count = count_bidi_control(input)
  local fw_in_ext = count_fullwidth_from(input, ext_start)
  local ext_in_ext = count_extend_from(input, ext_start)

  local classify
  -- Priority 1: any bidi format-control.
  local bidi_pos, bidi_cp = first_bidi_control(input)
  if bidi_pos ~= nil then
    classify = {
      kind = "Hazard",
      sub = { tag = "RloFlip", position = bidi_pos, control_cp = bidi_cp },
      positions = { bidi_pos },
      decoded = {},
    }
  else
    -- Priority 2: fullwidth/halfwidth in the extension.
    local fw_pos, fw_cp = first_fullwidth_from(input, ext_start)
    if fw_pos ~= nil then
      classify = {
        kind = "Hazard",
        sub = { tag = "WidthClassExt", position = fw_pos, cp = fw_cp },
        positions = { fw_pos },
        decoded = {},
      }
    else
      -- Priority 3: combining mark in the extension.
      local ext_pos, ext_cp = first_extend_from(input, ext_start)
      if ext_pos ~= nil then
        classify = {
          kind = "Hazard",
          sub = { tag = "CombiningInExt", position = ext_pos, cp = ext_cp },
          positions = { ext_pos },
          decoded = {},
        }
      elseif #dots >= M.MIN_MULTI_EXT then
        -- Priority 4: three or more extensions (advisory).
        local dots_copy = { unpack(dots) }
        classify = {
          kind = "Hazard",
          sub = { tag = "MultipleExtensions", dot_count = #dots },
          positions = dots_copy,
          decoded = {},
        }
      else
        classify = { kind = "Clear", sub = nil, positions = {}, decoded = {} }
      end
    end
  end

  local input_copy = {}
  if #input > 0 then
    input_copy = { unpack(input) }
  end

  return {
    input = input_copy,
    classify = classify,
    dot_positions = dots,
    last_dot_pos = last_dot,
    bidi_control_count = bidi_count,
    fullwidth_in_ext = fw_in_ext,
    combining_in_ext = ext_in_ext,
  }
end

-- True iff the verdict's classification is Clear.
function M.is_clear(verdict)
  return verdict.classify.kind == "Clear"
end

-- Human-facing tag for a verdict's classification, or nil when clear. Exhaustive
-- over the four sub-threats; an unknown tag is a construction bug and raises.
function M.classification_tag(verdict)
  local sub = verdict.classify.sub
  if sub == nil then
    return nil
  end
  local tag = sub.tag
  if tag == "RloFlip" then
    return "RloFlip"
  elseif tag == "WidthClassExt" then
    return "WidthClassExt"
  elseif tag == "CombiningInExt" then
    return "CombiningInExt"
  elseif tag == "MultipleExtensions" then
    return "MultipleExtensions"
  else
    error("classification_tag: unknown FilenameDisguise sub-threat " .. tostring(tag))
  end
end

-- Implicated positions (empty when clear).
function M.classification_positions(verdict)
  return verdict.classify.positions
end

return M

File: ports/elixir/lib/unicode_security/display/filename_disguise.ex
defmodule UnicodeSecurity.Display.FilenameDisguise do
  @moduledoc """
  filename-disguise — detection of filename/extension disguise attacks where the
  visible extension differs from the byte extension (the display-layer detector
  D).

  Byte-faithful transliteration of the verified rust reference implementation,
  itself a transcription of `Unicode/Security/Display/FilenameDisguise.lean`.

  Threat model. An adversary delivers a file whose rendered name looks like a
  benign type (`document.txt`) but whose actual byte extension is executable —
  the canonical attack inserts U+202E RIGHT-TO-LEFT OVERRIDE so
  `document<RLO>txt.exe` renders as `document exe.txt`.

  What the detector draws. Detection is presentation- and language-agnostic: it
  surfaces every codepoint that could cause display-vs-byte divergence in the
  filename — any purposeless bidi format-control anywhere (`BidiControlPurpose`:
  unbalanced, or a balanced span enclosing nothing right-to-left in a
  left-to-right context), and any fullwidth/halfwidth or combining (grapheme
  Extend) codepoint in the extension region (after the last `.`). Native-RTL
  names with no bidi controls clear, and so does a balanced embedding around an
  Arabic segment.

  Sub-threats (priority order):
    1. `RloFlip`            a purposeless bidi format-control in the input.
    2. `WidthClassExt`      a fullwidth/halfwidth codepoint in the extension.
    3. `CombiningInExt`     a combining (Extend) codepoint in the extension.
    4. `MultipleExtensions` >= 3 dots (advisory; e.g. legitimate `.tar.gz.sig`).

  It reuses the port's own tables — the bidi purpose rule, the Grapheme
  segmentation `Grapheme_Cluster_Break = Extend` class, and the inlined
  fullwidth range — never a host filesystem or rendering library.
  """

  alias UnicodeSecurity.Covert.BidiControlBalance
  alias UnicodeSecurity.Display.BidiControlPurpose
  alias UnicodeSecurity.Segmentation.Grapheme

  # ───────────────────────────────────────────────────────────────────
  # §1 Constants
  # ───────────────────────────────────────────────────────────────────

  # The count of `.` separators at or beyond which the name is treated as a
  # multiple-extension advisory hazard.
  @min_multi_ext 3

  @doc "The multiple-extension advisory threshold (`MIN_MULTI_EXT`)."
  def min_multi_ext, do: @min_multi_ext

  # ───────────────────────────────────────────────────────────────────
  # §2 Classification tags
  # ───────────────────────────────────────────────────────────────────

  @doc "Fixture-row tag string for a sub-threat map (matches `SubThreat.tag`)."
  def sub_threat_tag(%{kind: :rlo_flip}), do: "RloFlip"
  def sub_threat_tag(%{kind: :width_class_ext}), do: "WidthClassExt"
  def sub_threat_tag(%{kind: :combining_in_ext}), do: "CombiningInExt"
  def sub_threat_tag(%{kind: :multiple_extensions}), do: "MultipleExtensions"

  @doc "True iff the classification is `Clear`."
  def is_clear(%{kind: :clear}), do: true
  def is_clear(%{kind: :hazard}), do: false

  @doc "Human-facing tag for a hazard classification, or `nil` when clear."
  def classification_tag(%{kind: :clear}), do: nil
  def classification_tag(%{kind: :hazard, sub: sub}), do: sub_threat_tag(sub)

  @doc "Implicated codepoint positions of a classification (empty when clear)."
  def classification_positions(%{kind: :clear}), do: []
  def classification_positions(%{kind: :hazard, positions: positions}), do: positions

  # ───────────────────────────────────────────────────────────────────
  # §3 Core predicates (reuse the port's own tables)
  # ───────────────────────────────────────────────────────────────────

  @doc "True iff `cp` is U+002E FULL STOP (the extension separator)."
  def is_ascii_dot(cp), do: cp == 0x002E

  @doc "True iff `cp` is in the Halfwidth/Fullwidth Forms block."
  def is_fullwidth_halfwidth(cp), do: cp >= 0xFF01 and cp <= 0xFFEF

  @doc "True iff `cp` is a bidi format-control (reuses the port's own predicate)."
  def is_bidi_format_control(cp), do: BidiControlBalance.bidi_format_control?(cp)

  @doc "True iff `cp` has `Grapheme_Cluster_Break = Extend` (reuses the port's table)."
  def is_grapheme_extend(cp), do: Grapheme.lookup_gcb(cp) == :extend

  # ───────────────────────────────────────────────────────────────────
  # §4 Sub-detectors
  # ───────────────────────────────────────────────────────────────────

  # 0-based positions of every `.` in `input`.
  defp dot_positions(input) do
    input
    |> Enum.with_index()
    |> Enum.flat_map(fn {cp, idx} -> if is_ascii_dot(cp), do: [idx], else: [] end)
  end

  # Position and codepoint of the first purposeless bidi format-control, or
  # `nil`: unbalanced, or a balanced span enclosing nothing right-to-left in a
  # left-to-right context (`BidiControlPurpose.first_purposeless_control/1`). A
  # balanced embedding around an Arabic filename segment manages that segment
  # and is not a flip. Mirrors the Lean detect, which reads
  # `firstPurposelessControl`.
  defp first_bidi_control(input), do: BidiControlPurpose.first_purposeless_control(input)

  # Position and codepoint of the first fullwidth/halfwidth codepoint at or after
  # `start`, or `nil`.
  defp first_fullwidth_from(input, start) do
    input
    |> Enum.with_index()
    |> Enum.find_value(fn {cp, idx} ->
      if idx >= start and is_fullwidth_halfwidth(cp), do: {idx, cp}, else: nil
    end)
  end

  # Position and codepoint of the first Extend codepoint at or after `start`, or
  # `nil`.
  defp first_extend_from(input, start) do
    input
    |> Enum.with_index()
    |> Enum.find_value(fn {cp, idx} ->
      if idx >= start and is_grapheme_extend(cp), do: {idx, cp}, else: nil
    end)
  end

  # Count of fullwidth/halfwidth codepoints at or after `start`.
  defp count_fullwidth_from(input, start) do
    input
    |> Enum.with_index()
    |> Enum.count(fn {cp, idx} -> idx >= start and is_fullwidth_halfwidth(cp) end)
  end

  # Count of Extend codepoints at or after `start`.
  defp count_extend_from(input, start) do
    input
    |> Enum.with_index()
    |> Enum.count(fn {cp, idx} -> idx >= start and is_grapheme_extend(cp) end)
  end

  # ───────────────────────────────────────────────────────────────────
  # §5 Top-level detection
  # ───────────────────────────────────────────────────────────────────

  @doc """
  The FilenameDisguise detection function, reading its input as one filename.
  Mirrors the Lean detect, which is detectWithContext at the default context.
  """
  def detect(input), do: detect_with_context(false, input)

  @doc """
  The FilenameDisguise detection function under an explicit field context.
  Returns a verdict map mirroring the Lean/rust `Verdict`: `input`, `classify`,
  `dot_positions`, `last_dot_pos`, `bidi_control_count`, `fullwidth_in_ext`, and
  `combining_in_ext`. `running_text` mirrors the Lean `Context.runningText`: the
  extension rungs read the text after the last dot as a file extension, which a
  source file or a message does not have, so they do not run on running text;
  the purposeless-bidi-control rung holds of any field.
  """
  def detect_with_context(running_text, input) do
    dots = dot_positions(input)
    last_dot = List.last(dots)
    ext_start = if last_dot == nil, do: length(input), else: last_dot + 1

    bidi_count = Enum.count(input, &is_bidi_format_control/1)
    fw_in_ext = count_fullwidth_from(input, ext_start)
    ext_in_ext = count_extend_from(input, ext_start)

    classify = classify(input, dots, ext_start, running_text)

    %{
      input: input,
      classify: classify,
      dot_positions: dots,
      last_dot_pos: last_dot,
      bidi_control_count: bidi_count,
      fullwidth_in_ext: fw_in_ext,
      combining_in_ext: ext_in_ext
    }
  end

  # The priority ladder. The first trigger in priority order wins; when none
  # fires the input is `Clear`.
  defp classify(input, dots, ext_start, running_text) do
    cond do
      # Priority 1: a purposeless bidi format-control anywhere in the input.
      ctl = first_bidi_control(input) ->
        {pos, cp} = ctl
        hazard(%{kind: :rlo_flip, position: pos, control_cp: cp}, [pos])

      # The remaining rungs read an extension; running text has none.
      running_text ->
        %{kind: :clear}

      # Priority 2: fullwidth/halfwidth in the extension.
      fw = first_fullwidth_from(input, ext_start) ->
        {pos, cp} = fw
        hazard(%{kind: :width_class_ext, position: pos, cp: cp}, [pos])

      # Priority 3: combining mark in the extension.
      ext = first_extend_from(input, ext_start) ->
        {pos, cp} = ext
        hazard(%{kind: :combining_in_ext, position: pos, cp: cp}, [pos])

      # Priority 4: three or more extensions (advisory).
      length(dots) >= @min_multi_ext ->
        hazard(%{kind: :multiple_extensions, dot_count: length(dots)}, dots)

      # Otherwise the filename presents no display-vs-byte divergence trigger.
      true ->
        %{kind: :clear}
    end
  end

  defp hazard(sub, positions), do: %{kind: :hazard, sub: sub, positions: positions, decoded: []}
end
